File: frontend/src/types/appointment.ts
/**
 * Tipos para Appointments (Turnos)
 */

export enum AppointmentStatus {
  SCHEDULED = 'scheduled',
  CONFIRMED = 'confirmed',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
  NO_SHOW = 'no_show'
}

export interface Appointment {
  id: number;
  consultorio_id: number;
  patient_id: number;
  doctor_id: number;
  start_time: string; // ISO 8601 datetime
  end_time: string;
  reason?: string;
  notes?: string;
  status: AppointmentStatus;
  created_at: string;
  updated_at: string;
  // Datos anidados del paciente y doctor
  patient_name?: string;
  patient_phone?: string;
  doctor_name?: string;
}

export interface AppointmentCreate {
  patient_id: number;
  doctor_id: number;
  start_time: string;
  end_time: string;
  reason?: string;
  notes?: string;
  status?: AppointmentStatus;
}

export interface AppointmentUpdate {
  patient_id?: number;
  doctor_id?: number;
  start_time?: string;
  end_time?: string;
  reason?: string;
  notes?: string;
  status?: AppointmentStatus;
}

export interface AppointmentFilters {
  doctor_id?: number;
  patient_id?: number;
  status?: AppointmentStatus;
  start_date?: string; // YYYY-MM-DD
  end_date?: string;
  skip?: number;
  limit?: number;
}

export interface AppointmentListResponse {
  appointments: Appointment[];
  total: number;
}

File: backend/app/schemas/appointment.py
"""
Schemas para Appointments (Turnos)
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, field_validator
from app.models.appointment import AppointmentStatus


# ============= BASE SCHEMAS =============

class AppointmentBase(BaseModel):
    """Schema base para appointments"""
    patient_id: int
    doctor_id: int
    start_time: datetime
    end_time: datetime
    reason: Optional[str] = None
    notes: Optional[str] = None
    status: AppointmentStatus = AppointmentStatus.SCHEDULED


# ============= REQUEST SCHEMAS =============

class AppointmentCreate(AppointmentBase):
    """Schema para crear un turno"""
    
    @field_validator('end_time')
    @classmethod
    def validate_end_time(cls, v, info):
        """Validar que end_time sea después de start_time"""
        if 'start_time' in info.data and v <= info.data['start_time']:
            raise ValueError('end_time debe ser posterior a start_time')
        return v


class AppointmentUpdate(BaseModel):
    """Schema para actualizar un turno"""
    patient_id: Optional[int] = None
    doctor_id: Optional[int] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    reason: Optional[str] = None
    notes: Optional[str] = None
    status: Optional[AppointmentStatus] = None


class AppointmentStatusUpdate(BaseModel):
    """Schema para actualizar solo el estado"""
    status: AppointmentStatus


# ============= RESPONSE SCHEMAS =============

class AppointmentResponse(AppointmentBase):
    """Schema de respuesta con datos completos"""
    id: int
    consultorio_id: int
    created_at: datetime
    updated_at: datetime
    
    # Datos del paciente (nested)
    patient_name: Optional[str] = None
    patient_phone: Optional[str] = None
    
    # Datos del doctor (nested)
    doctor_name: Optional[str] = None

    class Config:
        from_attributes = True


class AppointmentListResponse(BaseModel):
    """Schema para lista de turnos"""
    appointments: list[AppointmentResponse]
    total: int
